experiments/process_sandbox/src/library_runner.cc

```diff
@@ -466,6 +467,134 @@ namespace
     return {response.response, std::move(in_fd)};
   }
 
+  /**
+   * Argument direction: for pointers, are they used to transfer data to or
+   * from the callee?
+   */
+  enum ArgDirection
+  {
+    /**
+     * The callee will read via this pointer.
+     */
+    In,
+    /**
+     * The callee will write via this pointer.
+     */
+    Out,
+    /**
+     * The callee will read and write via this pointer.
+     */
+    InOut
+  };
+
+  /**
+   * Wrapper for pointer arguments.  Encapsulates a pointer to a `T` that is to
+   * be passed to the parent.  Checks that the argument is inside the region
+   * that is shared with the parent, allocating a new buffer if it is not.  If
+   * this is an `Out` or `InOut` parameter, copies the result back.
+   *
+   * TODO: Handle arrays
+   */
+  template<typename T, ArgDirection Direction = In>
+  class CallbackPtrArg
+  {
+    /**
+     * The pointer captured by this object.
+     */
+    T* original;
+    /**
+     * An on-heap pointer for a copy of the pointee, allocated in the shared
+     * region.  If `original` is in the shared heap, this will be `nullptr`.
+     */
+    std::unique_ptr<T> copy;
+
+  public:
+    /**
+     * Constructor, captures `orig` and also allocates a copy of it on the
+     * heap.  The value of `orig` must not be modified for the lifetime of this
+     * object.
```

Comment:
This is not a type that's exposed to the library, if a multithreaded sandboxed library wants to race the thread using this class, it will send some nonsense data to the parent, but that's its problem.  The parent will always defensively copy this to protect against TOCTOU problems. 
 If you allocate this somewhere that's not in the sandbox, this copies it, and then you modify the original, then the changes that you've made won't be included in the RPC.  If it's an in-out parameter, then your changes will be clobbered on return.

Reply:
Perfect, thanks!